### ISSUE TITLE
array size mismatch when using seismology.estimate_numax()

### ISSUE THREAD

**Issue by darthoctopus** [KeplerGO/lightkurve#774]
#### Problem description

For some combinations of values for the keyword arguments window_width and spacing supplied to the acf2d function, estimate_numax() fails with an array size mismatch error. This seems to be because the preallocated array size for the output of utils.autocorrelate does not actually match the size of the array returned by it.

In particular, this seems to happen for red giants observed at short cadence. While I can get the acf2d method to work by masking out frequencies above an (artificial) maximum frequency, this seems like a more subtle issue: just because the code doesn't break doesn't mean it's necessarily not otherwise faulty.

#### Example

```python
import numpy as np
from astropy import units as u
import lightkurve as lk

def get_PG():
        
    data = np.loadtxt("PSD_all.pow")
    ν = data[..., 0] * u.uHz
    P = u.Quantity(data[..., -1])
    
#     m = ν.value < 500
    m = slice(None, None)
    
    return lk.Periodogram(ν[m], P[m])

pg = get_PG()
seis = pg.flatten().to_seismology()
seis.estimate_numax(window_width=25, spacing=1)
```

For the provided power spectrum (see here: [PSD_all.zip](https://github.com/KeplerGO/lightkurve/files/4882335/PSD_all.zip)), this fails with spacing < 1.49945671858..., which I don't have a good explanation for:

```
---------------------------------------------------------------------------
ValueError                                Traceback (most recent call last)
<ipython-input-78-76ccb8c22ce8> in <module>
----> 1 seis.estimate_numax(window_width=25, spacing=1.499456718580)

~/.local/lib/python3.8/site-packages/lightkurve/seismology/core.py in estimate_numax(self, method, **kwargs)
    544         if method == "acf2d":
    545             from .numax_estimators import estimate_numax_acf2d
--> 546             result = estimate_numax_acf2d(self.periodogram, **kwargs)
    547         self.numax = result
    548         return result

~/.local/lib/python3.8/site-packages/lightkurve/seismology/numax_estimators.py in estimate_numax_acf2d(periodogram, numaxs, window_width, spacing)
    150     for idx, numax in enumerate(numaxs):
    151         acf = utils.autocorrelate(periodogram, numax, window_width=window_width, frequency_spacing=fs)      #Return the acf at this numax
--> 152         acf2d[:, idx] = acf                                     #Store the 2D acf
    153         metric[idx] = (np.sum(np.abs(acf)) - 1 ) / len(acf)  #Store the max acf power normalised by the length
    154 

ValueError: could not broadcast input array from shape (554) into shape (584)
```

(the line numbers don't match up because I added a bunch of print statements to see what's going on)

#### Expected behavior

It should not return a ValueError.

#### Environment

-  Linux (Arch)
-  lightkurve version 1.11.1
-  installed with pip

**Comment by barentsen** [comment]
Thanks for reporting this @darthoctopus!

@ojhall94 Would you be willing to have a look?

**Comment by ojhall94** [comment]
Thanks for raising this @darthoctopus! I had never considered the possibility of somebody using SC data to look at a giant.

This is a weird error indeed. When lowering the maximum frequency of the spectrum to below 500, `seismology` appears to estimate a numax and deltanu that are perfectly reasonable, and the code behaves as I would expect it to. I don't think there are any issues there. If the power excess is visible, I would also generally recommend cutting up your power spectrum to the region around the excess, to avoid wasting time calculating the ACF in regions that aren't of interest.

It looks like the error is due to a rounding issue when calculating the window over which the ACF is computed around the maximum frequency in the spectrum. The code should be robust against this, so I'm not sure why its happening. I'll try and look at it with fresh eyes tomorrow and produce a fix :)

**Comment by darthoctopus** [comment]
Hi @ojhall94,

I'm using the power spectrum disseminated by Tiago Campante as part of the analysis/modelling call for ι Dra (TIC 165722603), and I don't think it's the same one as on MAST.

I agree that it's likely to be a data issue. My first clue for this was that when I eventually got around to feeding it to pbjam, it got spat right back out with an assertion failure claiming that the frequencies are not uniformly spaced. Looking more closely at the raw text file, it seems that the frequency values are formatted in a pretty strange manner (with precision set by a fixed number of significant figures). While technically correct (from the pedantic presentation point of view) this however also means that the local difference in frequency exhibits dithering error:
![image](https://user-images.githubusercontent.com/5761408/87614484-21c7c100-c6de-11ea-9c4d-9e02c8bf1431.png)
The magnitude of these dithering errors (0.1%) seems about right for the extent of the array size mismatch. As you can see, the dithering error increases with frequency (or more accurately, as the number of decimal places decreases to maintain the same number of significant figures), which probably explains why the error doesn't appear when I mask out frequencies above 500 μHz, or when the window size is large.

I don't know if there are any good solutions to this. Perhaps the easiest way out would be simply to assert (as in pbjam) that the the input frequencies be uniformly spaced.

**Comment by ojhall94** [comment]
Hey @darthoctopus!

Glad we solved that mystery. I've made a PR (#780) that raises a ValueError if somebody tries to use the ACF2D estimators on a periodogram that isn't uniformly sampled, so that if this happens again its at least clear why! Under this new PR you won't be able to run any ACF2D operations even when only looking at frequencies below 280 uHz (seems the dithering is pretty large even there), although remeshing the frequency axis does solve this, as you proposed.

>(albeit returning a very different value vs. when only using points with ν < 500 μHz — but that's maybe a separate issue).

This is because the estimated numax is based on a smoothed version of the Autocorrelation metric, and this will be smoothed more for larger arrays. If you focus the area of study (ie to <280 uhz) you should get a more accurate/precise estimate :) Don't think this counts as a bug as the behaviour of the smoothing is documented, although I admit it isn't super transparent.

This issue will close once #780 is merged in, if you're happy!

**Comment by darthoctopus** [comment]
This seems a satisfactory resolution to me. Thanks for helping out!

**Comment by barentsen** [comment]
A solution was merged in #780.

Thank you @darthoctopus for opening the issue, and thank you @ojhall94 for fixing!! :+1: :+1:
